src/audio/module_adapter/module/cadence.c

```diff
@@ -490,6 +490,15 @@ static int cadence_codec_prepare(struct processing_module *mod)
 	struct module_data *codec = &mod->priv;
 	struct cadence_codec_data *cd = codec->private;
 
+	if (mod->priv.state >= MODULE_IDLE) {
+		/* TODO: handle the case that stream_params was changed afterwards (different from
+		 *       the one applied on module codec initialization).
+		 */
+		comp_dbg(dev, "cadence_codec_prepare(): skipped by module under state %d",
+			 mod->priv.state);
+		return 0;
+	}
+
 	comp_dbg(dev, "cadence_codec_prepare() start");
```

Comment:
I think we should merge this patch once the `TODO` is implemented. 

First observation is that we need to add the `if` above after the call to `cadence_codec_post_init`. For example if the codec id changes we won't see that because the codec is resolved inside `cadence_codec_post_init`.

Reply:
I am fine with that, although this patch should cause no behavior change as it is.
Who should be in charge of the implementation?

Reply:
@johnylin76 I assume since you sent this patch then there is an use case you want to support, right? 

Cc: @cujomalainey

Reply:
@dbaluta I don't see it that way.

This patch won't change the current behavior. The module specific prepare() will be called while in MODULE_IDLE state, and early returned by all existing modules without doing anything.

This patch is to provide the flow for modules to handle re-configuration for stream params change, which is not handled (even not informed) now. There are multiple ways to handle and it's better to be decided and implemented by the module owners, since they have the deeper knowledge for the processing in modules.

Reply:
Ok, I agree. we can merge this now.

Reply:
@dbaluta to clear things up, the issue is if we change pipe params, and dsp does not suspend between params calls, the result is the codec does not get new information and is misconfigured at playback and results in flow rate errors. We are not sure why this is designed this way, but we are pretty sure it is incorrect.